src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,294 @@
+
+
+class LinksFile(JsonFile):
+    """
+    A file describing the links between entities in a subgraph.
+    """
+
+    def __init__(self, file_name: str, content: JSON):
+        file_id, version, project_id = super()._parse_file_name(file_name)
+        super().__init__(file_id, version, file_name, content)
+        self.project_id = project_id
+
+
+class MetadataFile(JsonFile):
+    """
+    A file describing one entity (e.g. biomaterial, protocol) in an subgraph.
+    """
+
+    def __init__(self, file_name: str, content: JSON):
+        file_id, version = super()._parse_file_name(file_name)
+        assert file_id == content['provenance']['document_id'], file_name
+        super().__init__(file_id, version, file_name, content)
+
+
+class DescriptorFile(JsonFile):
+    """
+    A file containing the checksums and other information for asserting the
+    integrity of a data file.
+    """
+
+    def __init__(self, file_name: str, content: JSON):
+        file_id, version = super()._parse_file_name(file_name)
+        super().__init__(file_id, version, file_name, content)
+
+    @property
+    def to_manifest_entry(self):
+        """
+        The content of a descriptor transformed into a format ready to create a
+        ManifestEntry object.
+        """
+        return {
+            'content-type': self.content['content_type'],
+            'crc32c': self.content['crc32c'],
+            'indexed': True,
+            'name': self.content['file_name'],
+            's3_etag': self.content['s3_etag'],
+            'sha1': self.content['sha1'],
+            'sha256': self.content['sha256'],
+            'size': self.content['size'],
+            'uuid': self.content['file_id'],
+            'version': self.content['file_version']
+        }
+
+
+@dataclass
+class StagingArea:
+    links: Mapping[str, LinksFile]
+    metadata: Mapping[str, MetadataFile]
+    descriptors: Mapping[str, DescriptorFile]
+
+    def file_ids_by_entity_type(self,
+                                subgraph_id: str
+                                ) -> Mapping[str, Set[str]]:
+        """
+        Parse the links in a subgraph and return a mapping of the metadata file
+        types (e.g. 'analysis_file', 'cell_suspension') to a set of file IDs
+        """
+        links_file: LinksFile = self.links[subgraph_id]
+        links_json = links_file.content
+        file_ids = defaultdict(set)
+        # Project ID is only mentioned in the links JSON if there is a
+        # supplementary_file_link so add it in here to make sure it is included.
+        file_ids['project'].add(links_file.project_id)
+        link: JSON
+        for link in links_json['links']:
+            link_type = link['link_type']
+            if link_type == 'process_link':
+                file_type = link['process_type']
+                file_id = link['process_id']
+                file_ids[file_type].add(file_id)
+                for process_link_type in ('input', 'output', 'protocol'):
```

Comment:
```suggestion
                for category in ('input', 'output', 'protocol'):
```

https://github.com/DataBiosphere/azul/commit/00bf40e6b1005642410c0f303032a895ee56dd4d#diff-dce1aec57757304e7f56e949fd1b69f0725a3ea33ac127fa3772baff1c4305ccR127